### ISSUE TITLE
proper:check_spec passing on Erlang R16&17, but not 18

### ISSUE THREAD

**Issue by danikp** [manopapad/proper#103]
same tests not passing on Erlang 18 even with latest proper, but passing on previous Erlang version
for example travis-ci build 
https://travis-ci.org/danikp/ChicagoBoss/builds/83380049

code is here
https://github.com/danikp/ChicagoBoss/blob/tests/test/boss_controller_compiler_test.erl#L8

**Comment by kostis** [comment]
This is not a PropEr issue.

Instead, it's an issue of the code of Chicago Boss.  Its specs are wrong. For example, this one: https://github.com/danikp/ChicagoBoss/blob/master/src/boss/boss_controller_compiler.erl#L53
Its return type should be something which is a list() not a syntaxTree().

What happened is that the types of erl_parse where strengthened in 18.x compared to those in R16* and 17.x.  Consequently, these erroneous specs are now found as erroneous by PropEr when they were not before.  You better fix them.  If you start doing this, you may realize that the declared types of erl_syntax.erl are perhaps too strict and may need to be relaxed.

**Comment by kostis** [comment]
Contrary to what I wrote above, PropEr had indeed an issue with the handling of user-defined types in 18.x when checking specs. This commit (https://github.com/manopapad/proper/commit/ced2ddb3a719aec22e5abe55c51b56658b0e8a87) fixes it.

But what I wrote above is, I believe, at least partly correct.  Some of the specs of Chicago Boss probably need some fixing.

Thanks for your report and let us know if you encounter any more issues.
